In [2]:
import pandas as pd
import numpy as np

In [2]:
#df = pd.read_csv('data/Train1.csv', error_bad_lines=False, index_col=False, dtype= 'unicode')


In [3]:
df = pd.read_csv('data/IST_corrected.csv', low_memory = False)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19435 entries, 0 to 19434
Columns: 112 entries, HOSPNUM to NCB14
dtypes: float64(20), int64(29), object(63)
memory usage: 16.6+ MB


In [4]:
df.head()

ID HOSPNUM RDELAY RCONSC SEX AGE RSLEEP RATRIAL RCT RVISINF  ... DEAD8 H14  \
0  1       1     17      D   M  69      Y     NaN   Y       Y  ...     0   0   
1  2       1     10      F   M  76      Y     NaN   Y       N  ...     0   0   
2  3       1     24      F   M  23      N     NaN   Y       N  ...     0   0   
3  4       1      5      D   F  83      N     NaN   N       N  ...     0   0   
4  6       1      8      D   F  64      Y     NaN   Y       Y  ...     0   0   

  ISC14 NK14 STRK14 HTI14 PE14 DVT14 TRAN14 NCB14  
0     0    0      0     0    0     0      0     0  
1     0    0      0     0    0     0      0     0  
2     0    0      0     0    0     0      0     0  
3     0    0      0     0    0     0      0     0  
4     0    0      0     0    0     0      0     0  

[5 rows x 113 columns]

In [155]:
#df.columns.tolist()

In [136]:
#df.isnull().sum().tolist()

In [5]:
# check for duplicates
double = df[df.duplicated()]
double

Empty DataFrame
Columns: [HOSPNUM, RDELAY, RCONSC, SEX, AGE, RSLEEP, RATRIAL, RCT, RVISINF, RHEP24, RASP3, RSBP, RDEF1, RDEF2, RDEF3, RDEF4, RDEF5, RDEF6, RDEF7, RDEF8, STYPE, RDATE, HOURLOCAL, MINLOCAL, DAYLOCAL, RXASP, RXHEP, DASP14, DASPLT, DLH14, DMH14, DHH14, ONDRUG, DSCH, DIVH, DAP, DOAC, DGORM, DSTER, DCAA, DHAEMD, DCAREND, DTHROMB, DMAJNCH, DMAJNCHD, DMAJNCHX, DSIDE, DSIDED, DSIDEX, DDIAGISC, DDIAGHA, DDIAGUN, DNOSTRK, DNOSTRKX, DRSISC, DRSISCD, DRSH, DRSHD, DRSUNK, DRSUNKD, DPE, DPED, DALIVE, DALIVED, DPLACE, DDEAD, DDEADD, DDEADC, DDEADX, FDEAD, FLASTD, FDEADD, FDEADC, FDEADX, FRECOVER, FDENNIS, FPLACE, FAP, FOAC, FU1_RECD, FU2_DONE, COUNTRY, CNTRYNUM, FU1_COMP, NCCODE, CMPLASP, CMPLHEP, DIED, TD, EXPDD, EXPD6, EXPD14, SET14D, ID14, OCCODE, DEAD1, DEAD2, DEAD3, DEAD4, DEAD5, ...]
Index: []

[0 rows x 112 columns]

In [60]:
df_types = df.columns.to_series().groupby(df.dtypes).groups
{k.name: v for k, v in df_types.items()}

{'int64': Index(['HOSPNUM', 'RDELAY', 'AGE', 'RSBP', 'HOURLOCAL', 'MINLOCAL', 'DAYLOCAL',
        'CNTRYNUM', 'DIED', 'SET14D', 'ID14', 'OCCODE', 'DEAD1', 'DEAD2',
        'DEAD3', 'DEAD4', 'DEAD5', 'DEAD6', 'DEAD7', 'DEAD8', 'H14', 'ISC14',
        'NK14', 'STRK14', 'HTI14', 'PE14', 'DVT14', 'TRAN14', 'NCB14'],
       dtype='object'),
 'float64': Index(['ONDRUG', 'DMAJNCHD', 'DSIDED', 'DRSISCD', 'DRSHD', 'DRSUNKD', 'DPED',
        'DALIVED', 'DDEADD', 'DDEADC', 'FLASTD', 'FDEADD', 'FDEADC', 'FU1_RECD',
        'FU2_DONE', 'FU1_COMP', 'TD', 'EXPDD', 'EXPD6', 'EXPD14'],
       dtype='object'),
 'object': Index(['RCONSC', 'SEX', 'RSLEEP', 'RATRIAL', 'RCT', 'RVISINF', 'RHEP24',
        'RASP3', 'RDEF1', 'RDEF2', 'RDEF3', 'RDEF4', 'RDEF5', 'RDEF6', 'RDEF7',
        'RDEF8', 'STYPE', 'RDATE', 'RXASP', 'RXHEP', 'DASP14', 'DASPLT',
        'DLH14', 'DMH14', 'DHH14', 'DSCH', 'DIVH', 'DAP', 'DOAC', 'DGORM',
        'DSTER', 'DCAA', 'DHAEMD', 'DCAREND', 'DTHROMB', 'DMAJNCH', 'DMAJNCHX',
        

In [4]:
object_col = list(df.select_dtypes(include='object').columns)
num_col = list(df.select_dtypes(include=['float64', 'int64']).columns)

In [58]:
#object_col

In [5]:
# Iterate over object columns
for column in object_col:
    columnSeriesObj = df[column]
    print()
    print('Colunm Name : ', column)
    print('Describe: ', columnSeriesObj.describe())
    if columnSeriesObj.nunique() <= 6:
        df[column] = df[column].str.upper()                     # all strings in uppercase
        print('Unique values : ', columnSeriesObj.unique())


Colunm Name :  RCONSC
Describe:  count     19435
unique        3
top           F
freq      14921
Name: RCONSC, dtype: object
Unique values :  ['D' 'F' 'U']

Colunm Name :  SEX
Describe:  count     19435
unique        2
top           M
freq      10407
Name: SEX, dtype: object
Unique values :  ['M' 'F']

Colunm Name :  RSLEEP
Describe:  count     19435
unique        2
top           N
freq      13750
Name: RSLEEP, dtype: object
Unique values :  ['Y' 'N']

Colunm Name :  RATRIAL
Describe:  count     18451
unique        2
top           N
freq      15282
Name: RATRIAL, dtype: object
Unique values :  [nan 'Y' 'N']

Colunm Name :  RCT
Describe:  count     19435
unique        2
top           Y
freq      13024
Name: RCT, dtype: object
Unique values :  ['Y' 'N']

Colunm Name :  RVISINF
Describe:  count     19435
unique        2
top           N
freq      13020
Name: RVISINF, dtype: object
Unique values :  ['Y' 'N']

Colunm Name :  RHEP24
Describe:  count     19091
unique        2
top           N


In [6]:
# Iterate over numeric columns
for column in num_col:
    columnSeriesObj = df[column]
    print()
    print('Colunm Name : ', column)
    print('Describe: ', columnSeriesObj.describe())



Colunm Name :  HOSPNUM
Describe:  count    19435.000000
mean       200.055467
std        160.335093
min          1.000000
25%         60.000000
50%        159.000000
75%        319.000000
max        568.000000
Name: HOSPNUM, dtype: float64

Colunm Name :  RDELAY
Describe:  count    19435.000000
mean        20.124363
std         12.470330
min          1.000000
25%          9.000000
50%         19.000000
75%         29.000000
max         48.000000
Name: RDELAY, dtype: float64

Colunm Name :  AGE
Describe:  count    19435.000000
mean        71.715410
std         11.619714
min         16.000000
25%         65.000000
50%         73.000000
75%         80.000000
max         99.000000
Name: AGE, dtype: float64

Colunm Name :  RSBP
Describe:  count    19435.000000
mean       160.159197
std         27.610382
min         70.000000
25%        140.000000
50%        160.000000
75%        180.000000
max        295.000000
Name: RSBP, dtype: float64

Colunm Name :  HOURLOCAL
Describe:  count    19435.

In [8]:
df.RDATE.unique() # probably polish (https://en.wikipedia.org/wiki/Slavic_calendar)

array(['sty-91', 'lut-91', 'mar-91', 'kwi-91', 'maj-91', 'cze-91',
       'lip-91', 'sie-91', 'wrz-91', 'pa�-91', 'lis-91', 'gru-91',
       'sty-92', 'lut-92', 'mar-92', 'kwi-92', 'maj-92', 'cze-92',
       'lip-92', 'sie-92', 'wrz-92', 'pa�-92', 'lis-92', 'gru-92',
       'sty-93', 'lut-93', 'mar-93', 'lip-93', 'sty-94', 'wrz-93',
       'sie-93', 'gru-93', 'lis-93', 'kwi-93', 'mar-94', 'wrz-94',
       'lip-94', 'sie-94', 'maj-93', 'pa�-93', 'kwi-94', 'lut-94',
       'maj-94', 'cze-93', 'cze-94', 'gru-94', 'sty-95', 'pa�-94',
       'lis-94', 'lut-95', 'lip-95', 'maj-96', 'maj-95', 'mar-95',
       'wrz-95', 'pa�-95', 'kwi-95', 'lut-96', 'sty-96', 'cze-95',
       'sie-95', 'lis-95', 'gru-95', 'mar-96', 'kwi-96'], dtype=object)

In [9]:
#extract year from RDATE
df['RYEAR'] = df.RDATE.str.extract(r'(\d{2})', expand=False)

In [11]:
df.RYEAR.unique()

array(['91', '92', '93', '94', '95', '96'], dtype=object)

Month (EN) | Month (PL) | Month (in df)
:----- | ----: | -----:
January  | styczeń | sty
February  | luty  | lut
March | (brzezień) | Mar
April | kwiecień | kwi
May | (trawień) | maj
June | czerwiec | cze
July | lipiec | lip
August | sierpień | sie
September | wrzesień | wrz
Ocotber | październik | pa�
November | listopad | lis
December | grudzień | gru

In [12]:
#df.RDATE.to_string()
#type(df.RDATE)

In [10]:
# equivalents/translations for months
month_dict = {'sty': '01', 'lut': '02', 'mar': '03', 'kwi': '04', 'maj': '05', 'cze': '06',
                'lip': '07', 'sie': '08', 'wrz': '09', 'pa�': '10', 'lis': '11', 'gru': '12'}

In [11]:
# replace string in RDATE with number (according to month)
df.RDATE = df.RDATE.replace(month_dict, regex=True)

In [12]:
df.RDATE.unique()

array(['01-91', '02-91', '03-91', '04-91', '05-91', '06-91', '07-91',
       '08-91', '09-91', '10-91', '11-91', '12-91', '01-92', '02-92',
       '03-92', '04-92', '05-92', '06-92', '07-92', '08-92', '09-92',
       '10-92', '11-92', '12-92', '01-93', '02-93', '03-93', '07-93',
       '01-94', '09-93', '08-93', '12-93', '11-93', '04-93', '03-94',
       '09-94', '07-94', '08-94', '05-93', '10-93', '04-94', '02-94',
       '05-94', '06-93', '06-94', '12-94', '01-95', '10-94', '11-94',
       '02-95', '07-95', '05-96', '05-95', '03-95', '09-95', '10-95',
       '04-95', '02-96', '01-96', '06-95', '08-95', '11-95', '12-95',
       '03-96', '04-96'], dtype=object)

In [21]:
print(df.COUNTRY.value_counts())
df.CNTRYNUM.value_counts()

UK      3197
ITAL    1779
SWIT     854
POLA     364
NETH     344
SWED     310
AUSL     276
NORW     274
ARGE     263
SPAI     242
CZEC     234
NEW      231
PORT     205
BELG     144
TURK     141
AUST     112
INDI     102
GREE      66
SING      65
USA       61
CANA      54
ISRA      53
HUNG      49
HONG      49
SLOK      43
FINL      38
BRAS      37
EIRE      34
CHIL      31
SOUT      31
SLOV      27
DENM      20
ROMA      15
SRI       12
JAPA       4
FRAN       1
Name: COUNTRY, dtype: int64


27    3197
14    1779
25     854
18     364
15     344
24     310
1      276
17     274
29     263
22     242
7      234
16     231
19     205
3      144
35     141
2      112
37     102
31      66
34      65
28      61
5       54
13      53
36      49
30      49
44      43
10      38
42      37
9       34
6       31
21      31
20      27
8       20
33      15
23      12
38       4
11       1
Name: CNTRYNUM, dtype: int64

In [13]:
df['COUNTRY_ISO'] = df['COUNTRY'].replace({'UK': 'GB', 'ITAL': 'IT', 'SWIT': 'CH', 'POLA': 'PL', 'NETH': 'NL', 
    'SWED': 'SE', 'AUSL': 'AU', 'NORW': 'NO', 'ARGE': 'AR', 'SPAI': 'ES', 'CZEC': 'CZ', 'NEW': 'NZ', 'PORT': 'PT',
    'BELG': 'BE', 'TURK': 'TR', 'AUST': 'AT', 'INDI': 'IN', 'GREE': 'GR', 'SING': 'SG', 'USA': 'USA', 'CANA': 'CA',
    'ISRA': 'IL', 'HUNG': 'HU', 'HONG': 'HK', 'SLOK': 'SK', 'FINL': 'FI', 'BRAS': 'BR', 'EIRE': 'IE', 'CHIL': 'CL',
    'SOUT': 'ZA', 'SLOV': 'SI', 'DENM': 'DK', 'ROMA': 'RO', 'SRI': 'LK', 'JAPA': 'JP', 'FRAN': 'FR'})

In [165]:
#df.COUNTRY_ISO.value_counts()

In [14]:
pd.qcut(x=df['AGE'], q=5, precision=0)

0        (63.0, 70.0]
1        (70.0, 76.0]
2        (70.0, 76.0]
3        (76.0, 82.0]
4        (76.0, 82.0]
             ...     
19430    (63.0, 70.0]
19431    (70.0, 76.0]
19432    (76.0, 82.0]
19433    (82.0, 99.0]
19434    (15.0, 63.0]
Name: AGE, Length: 19435, dtype: category
Categories (5, interval[float64]): [(15.0, 63.0] < (63.0, 70.0] < (70.0, 76.0] < (76.0, 82.0] < (82.0, 99.0]]

In [15]:
# bin AGE into equally sized bins
labels = ['< 63', '< 71', '< 76', '< 82', '> 82']
df['AGE_bin'] = pd.qcut(x=df['AGE'], q=5, labels = labels, precision=0)

In [25]:
group_hep24 = df.groupby(by = ['RHEP24', 'RASP3'])
group_hep24['RHEP24'].value_counts()

RHEP24  RASP3  RHEP24
N       N      N         14195
        Y      N          3827
Y       N      Y           316
        Y      Y           113
Name: RHEP24, dtype: int64

In [16]:
# replace H with M
df['RXHEP'].replace('H','M', inplace=True)

In [34]:
print(df.RXHEP.describe())
print(df.RXHEP.unique())

count     19435
unique        3
top           N
freq       9718
Name: RXHEP, dtype: object
['N' 'L' 'M']


In [17]:
# merge DHH14 into DMH14 
# (DHH14 = Medium dose heparin given for 14 days etc in pilot (combine with DMH14))
df['DMH14'] = np.where(df['DMH14'].isnull() == True, df['DHH14'], df['DMH14'])

In [18]:
# new variable: RXHEP14
# Heparin given for 14 days or till death/discharge (Y, N)
#df['RXHEP14'] = np.where(df["RXHEP"] == 'M', df['DMH14'], 'N')
#df['RXHEP14'] = np.where(df["RXHEP"] == 'L', df['DLH14'], 'N')

conditions = [
    df['RXHEP'].eq('M') & df['DMH14'].eq('Y'),
    df['RXHEP'].eq('L') & df['DLH14'].eq('Y')]

choices = ['Y','Y']

df['RXHEP14'] = np.select(conditions, choices, default='N')

In [19]:
# new variable: HEP14
# Dose Heparin given for 14 days or till death/discharge (Y, N)

conditions = [
    df['RXHEP'].eq('M') & df['DMH14'].eq('Y'),
    df['RXHEP'].eq('L') & df['DLH14'].eq('Y')]

choices = ['M','L']

df['HEP14'] = np.select(conditions, choices, default='N')

In [20]:
# Do the patients receive the treatment they were assigned to? 
# No => Go with DASP14 an HEP14, i.e. the treatment the patients actually received. 
print(df['RXHEP'].groupby(by = [df['HEP14']]).value_counts())
print(df['RXASP'].groupby(df['DASP14']).value_counts())

HEP14  RXHEP
L      L        4450
M      M        4242
N      N        9718
       M         614
       L         411
Name: RXHEP, dtype: int64
DASP14  RXASP
N       N        9552
        Y         775
U       N           9
        Y           5
Y       Y        8929
        N         143
Name: RXASP, dtype: int64


In [29]:
# new variable: TREAT14 (Meds given for 14 days or till death or discharge)
# Treatment: 'Aspirin','Aspirin + Medium Heparin','Aspirin + Low Heparin', 'Medium Heparin', 'Low Heparin'

conditions = [
    df['DASP14'].eq('Y') & df['HEP14'].eq('N'),     # Aspirin - no Heparin
    df['DASP14'].eq('Y') & df['HEP14'].eq('M'),     # Aspirin + Medium Heparin
    df['DASP14'].eq('Y') & df['HEP14'].eq('L'),     # Aspirin + Low Heparin
    df['HEP14'].eq('M'),                            # Medium Heparin
    df['HEP14'].eq('L')]                            # Low Heparin

choices = ['Aspirin','Aspirin + Medium Heparin','Aspirin + Low Heparin', 'Medium Heparin', 'Low Heparin']

df['TREAT14'] = np.select(conditions, choices, default='Control')

In [84]:
df.TREAT14.value_counts()

Control                     2983
Aspirin                     2514
Low Heparin                 1136
Aspirin + Low Heparin       1098
Medium Heparin              1042
Aspirin + Medium Heparin     989
Name: TREAT14, dtype: int64

In [30]:
# new variable: DMEDS
# additional medication beside ASP or HEP

conditions = [
    df['DSCH'].eq('Y'), df['DIVH'].eq('Y'), df['DAP'].eq('Y'), df['DOAC'].eq('Y'),
    df['DGORM'].eq('Y'), df['DSTER'].eq('Y'), df['DCAA'].eq('Y')]

choices = ['Subcut. Heparin','Intrav. Heparin','Antiplatelet drug','Other coanticoagulants',
    'Glycerol or mannitol','Steroids','Calcium antagonist']

df['DMEDS'] = np.select(conditions, choices, default='N')

In [28]:
df.DMEDS.describe()

count     9762
unique       8
top          N
freq      7007
Name: DMEDS, dtype: object

In [31]:
# new variable: DIAG
# combine all types of first stroke

conditions = [
    df['DDIAGISC'].eq('Y'), df['DDIAGHA'].eq('Y'), df['DDIAGUN'].eq('Y'), df['DNOSTRK'].eq('Y')]

choices = ['Ischaemic stroke','Haemorrhagic stroke','Indeterminate stroke','Not a stroke']

df['DIAG'] = np.select(conditions, choices, default='Unknown')

In [32]:
# new variable: RDIAG14
# combine all types of recurrent stroke (within 14 days)

conditions = [
    df['DRSISC'].eq('Y'), df['DRSH'].eq('Y'), df['DRSUNK'].eq('Y'), df['DPE'].eq('Y')]

choices = ['Ischaemic stroke','Haemorrhagic stroke','Unknown type', 'Pulmonary embolism']

df['RDIAG14'] = np.select(conditions, choices, default='None')

In [46]:
df.RDIAG14.value_counts()

None                   9303
Ischaemic stroke        206
Unknown type            136
Pulmonary embolism       67
Haemorrhagic stroke      50
Name: RDIAG14, dtype: int64

In [37]:
from math import nan

In [52]:
# new variable: RDISCH14
# Discharged after 14 days - alive or dead?

conditions = [
    df['DALIVE'].eq('Y'), df['DDEAD'].eq('Y'), df['DALIVE'].eq('N')]

choices = ['Discharged alive from hospital', 'Dead on discharge form', 'Not discharged']

df['RDISCH14'] = np.select(conditions, choices, default= np.nan)

In [53]:
df.RDISCH14.value_counts()

Discharged alive from hospital    10322
Not discharged                     7055
Dead on discharge form             2034
nan                                  24
Name: RDISCH14, dtype: int64

In [54]:
df.RDISCH14.isnull().sum()

0

In [49]:
# neither dead nor alive?! - not discharged?
# DALIVE - Discharged alive from hospital
# DDEAD - Dead on discharge form
print(df['DDEAD'].groupby(df['DALIVE']).value_counts())

DALIVE  DDEAD
N       N        3516
        Y        1062
U       U           2
Y       N        5163
Name: DDEAD, dtype: int64


In [26]:
# DPLACE - Discharge destination
# only patients that were alive were discharged to certain destinations
print(df['DALIVE'].groupby(df['DPLACE']).value_counts())
print(df['DDEAD'].groupby(df['DPLACE']).value_counts())

DPLACE  DALIVE
A       Y         3553
B       Y          138
C       Y          118
D       Y          258
E       Y          787
U       Y            7
Name: DALIVE, dtype: int64
DPLACE  DDEAD
A       N        3553
B       N         138
C       N         118
D       N         258
E       N         787
U       N           7
Name: DDEAD, dtype: int64


In [43]:
# discharged patients were alive and may have died within 6 months
group_da = df.groupby(by=['DPLACE', 'DDEAD','DALIVE'])
group_da['FDEAD'].value_counts()

DPLACE  DDEAD  DALIVE  FDEAD
A       N      Y       N        3324
                       Y         219
B       N      Y       N         127
                       Y          11
C       N      Y       N          98
                       Y          20
D       N      Y       N         189
                       Y          69
E       N      Y       N         656
                       Y         128
U       N      Y       N           7
Name: FDEAD, dtype: int64

In [76]:
# DDEADC - Cause of death
# 0 in DDEADC have not died?
print(df['DDEAD'].groupby(df['DDEADC']).value_counts())
print(df['DALIVE'].groupby(df['DDEADC']).value_counts())

DDEADC  DDEAD
0       N        150
1       Y        526
2       Y        116
3       Y         25
4       Y        191
5       Y         71
6       Y         49
7       Y         61
8       Y         28
Name: DDEAD, dtype: int64
DDEADC  DALIVE
0       N          80
        Y          70
1       N         524
2       N         116
3       N          25
4       N         190
5       N          71
6       N          47
7       N          60
8       N          28
Name: DALIVE, dtype: int64


In [42]:
# when DALIVE = N equals "not discharged" then patients should appear in OCCODE (outcome @ 6 months; 1 - dead @ 6 months)
group_da = df.groupby(by=['OCCODE', 'DDEAD','DALIVE'])
group_da['FDEAD'].value_counts()

OCCODE  DDEAD  DALIVE  FDEAD
1       N      N       Y         684
               Y       Y         442
        Y      N       Y        1062
2       N      N       N        1900
                       Y          32
               Y       N        1928
                       Y          36
        U      U       N           2
3       N      N       N         572
               Y       N        1385
4       N      N       N         321
               Y       N        1359
Name: FDEAD, dtype: int64

In [87]:
# OCCODE has more info on dependency/recovery after 6 months than FDENNIS (no unknown)
group_depen = df.groupby(by=['OCCODE', 'FDENNIS'])
print(group_depen['SEX'].value_counts())

OCCODE  FDENNIS  SEX
1       N        F         1
2       U        F        19
                 M        10
        Y        F      3981
                 M      3746
3       N        M      2424
                 F      1440
4       N        M      2045
                 F      1245
        U        F         4
                 M         2
9       U        M        10
                 F         9
Name: SEX, dtype: int64


In [77]:

# new variable: DEAD
# Cause of death (regardless of time - within 14 days or 6 months)

conditions = [
    df['DEAD1'].eq(1), df['DEAD2'].eq(1), df['DEAD3'].eq(1), df['DEAD4'].eq(1), 
    df['DEAD5'].eq(1), df['DEAD6'].eq(1), df['DEAD7'].eq(1), df['DEAD8'].eq(1)]

choices = ['Initial stroke','Recurrent stroke (ischemiac/unknown)', 'Recurrent stroke (haemorraghic)', 'Pneumonia', 'Coronary heart disease', 'Pulmonary embolism', 'Other vascular or unknown', 'Non vascular']

df['DEAD'] = np.select(conditions, choices, default='Alive')

In [79]:
df.DEAD.value_counts()

Alive                                   15194
Initial stroke                           1481
Pneumonia                                 836
Recurrent stroke (ischemiac/unknown)      538
Other vascular or unknown                 537
Coronary heart disease                    371
Non vascular                              230
Pulmonary embolism                        171
Recurrent stroke (haemorraghic)            77
Name: DEAD, dtype: int64

In [56]:
df['DDEADC'].replace({1: 'Initial Stroke', 2: 'Recurrent stroke (ischaemic or unknown)', 3: 'Recurrent stroke (haemorrhagic)', 
4: 'Pneumonia', 5: 'Coronary heart disease', 6: 'Pulmonary embolism', 7: 'Other vascular or unknown', 8: 'Non-vascular', 0: 'Unknown'}, inplace=True)

In [57]:
df['FDEADC'].replace({1: 'Initial Stroke', 2: 'Recurrent stroke (ischaemic or unknown)', 3: 'Recurrent stroke (haemorrhagic)', 
4: 'Pneumonia', 5: 'Coronary heart disease', 6: 'Pulmonary embolism', 7: 'Other vascular or unknown', 8: 'Non-vascular', 0: 'Unknown'}, inplace=True)

In [80]:
# check if correct - 4 patients not correctly assigned - but FDEADC is in line with all deaths (either within 14 days or 6 months - use this variable)
group_dc = df.groupby(by = ['DDEADC','FDEADC'])
group_dc['DEAD'].value_counts()

DDEADC                                   FDEADC                                   DEAD                                
Coronary heart disease                   Coronary heart disease                   Coronary heart disease                  158
Initial Stroke                           Initial Stroke                           Initial stroke                          976
Non-vascular                             Non-vascular                             Non vascular                             52
                                                                                  Other vascular or unknown                 3
Other vascular or unknown                Other vascular or unknown                Other vascular or unknown               112
                                                                                  Alive                                     1
Pneumonia                                Pneumonia                                Pneumonia                               374

In [ ]:
df['DPLACE'].replace({'A': 'Home', 'B': 'Relatives home', 'C': 'Residential care', 'D': 'Nursing home', 'E': 'Other hospital departments', 'U': 'Unknown'}, inplace=True)

In [ ]:
df['FDEADC'].replace({1: 'Initial Stroke', 2: 'Recurrent stroke (ischaemic or unknown)', 3: 'Recurrent stroke (haemorrhagic)', 
4: 'Pneumonia', 5: 'Coronary heart disease', 6: 'Pulmonary embolism', 7: 'Other vascular or unknown', 8: 'Non-vascular', 0: 'Unknown'}, inplace=True)

In [ ]:
df['OCCODE'].replace({1: 'Dead', 2: 'Dependent', 3: 'Not recovered', 4: 'Recovered', 0: 'Unknown'}, inplace=True)

In [81]:
for column in df.columns:
    columnSeriesObj = df[column]
    if columnSeriesObj.isnull().sum() != 0:
        print('Column Name: ', column)
        print('Missing values : ', columnSeriesObj.isnull().sum() )

Column Name:  RATRIAL
Missing values :  984
Column Name:  RHEP24
Missing values :  344
Column Name:  RASP3
Missing values :  984
Column Name:  DASP14
Missing values :  22
Column Name:  DASPLT
Missing values :  147
Column Name:  DLH14
Missing values :  22
Column Name:  DMH14
Missing values :  22
Column Name:  DHH14
Missing values :  18451
Column Name:  ONDRUG
Missing values :  1
Column Name:  DSCH
Missing values :  305
Column Name:  DIVH
Missing values :  305
Column Name:  DAP
Missing values :  18
Column Name:  DOAC
Missing values :  18
Column Name:  DGORM
Missing values :  23
Column Name:  DSTER
Missing values :  28
Column Name:  DCAA
Missing values :  29
Column Name:  DHAEMD
Missing values :  28
Column Name:  DCAREND
Missing values :  1005
Column Name:  DTHROMB
Missing values :  315
Column Name:  DMAJNCH
Missing values :  14
Column Name:  DMAJNCHD
Missing values :  19286
Column Name:  DMAJNCHX
Missing values :  19290
Column Name:  DSIDE
Missing values :  14
Column Name:  DSIDED
Missin

+ Drop NAs:
    + RATRIAL not noted during trial phase (984 patients) - kick out for model if needed
    + RHEP24 and RASP3 (info on meds prior to first stroke) - kick out if needed/no additional info
    + DMH14 and DLH14 - kick out 
    
+ Drop entirely:
    + DHH14 - (was merged with DMH14)
    + FDENNIS - is included in OCCODE ('dependent' after 6 months)
    + 'DMAJNCHX', 'DSIDED', 'DSIDEX', 'DNOSTRKX', 'DDEADX', 'FLASTD', 'FDEADX', 'NCCODE' - (no additional info and many missing values)

In [82]:
# drop the following columns entirely
df = df.drop(['DHH14', 'FDENNIS', 'DMAJNCHX', 'DSIDED', 'DSIDEX', 'DNOSTRKX', 'DDEADX', 'FLASTD', 'FDEADX', 'NCCODE'], axis=1)

In [84]:
for column in df.columns:
    columnSeriesObj = df[column]
    if columnSeriesObj.isnull().sum() != 0:
        print('Column Name: ', column)
        print('Missing values : ', columnSeriesObj.isnull().sum() )

Column Name:  RATRIAL
Missing values :  984
Column Name:  RHEP24
Missing values :  344
Column Name:  RASP3
Missing values :  984
Column Name:  DASP14
Missing values :  22
Column Name:  DASPLT
Missing values :  147
Column Name:  DLH14
Missing values :  22
Column Name:  DMH14
Missing values :  22
Column Name:  ONDRUG
Missing values :  1
Column Name:  DSCH
Missing values :  305
Column Name:  DIVH
Missing values :  305
Column Name:  DAP
Missing values :  18
Column Name:  DOAC
Missing values :  18
Column Name:  DGORM
Missing values :  23
Column Name:  DSTER
Missing values :  28
Column Name:  DCAA
Missing values :  29
Column Name:  DHAEMD
Missing values :  28
Column Name:  DCAREND
Missing values :  1005
Column Name:  DTHROMB
Missing values :  315
Column Name:  DMAJNCH
Missing values :  14
Column Name:  DMAJNCHD
Missing values :  19286
Column Name:  DSIDE
Missing values :  14
Column Name:  DDIAGISC
Missing values :  23
Column Name:  DDIAGHA
Missing values :  26
Column Name:  DDIAGUN
Missing v

In [128]:
## leave for later
#df_dropped = df.dropna(subset=[])
#df_dropped.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9534 entries, 175 to 9761
Columns: 123 entries, ID to DEAD
dtypes: object(123)
memory usage: 9.0+ MB


In [85]:
#save clean file
df.to_csv('data/IST_corrected_clean.csv') 